csrc/fusion_segmenter.cpp

```diff
@@ -4206,6 +4207,124 @@ void SegmentCandidateFinder::findSegments() {
+        continue;
+      }
+
+      auto precisions =
+          ir_utils::getPrecisionOfProducerConsumerTensors(maybe_upcast_op);
+      if (!precisions.has_value() || precisions->first >= precisions->second) {
+        continue;
+      }
+
+      // Check if there's multiple uses of the upcast output
+      auto uses_of_upcast_out_tv = maybe_upcast_out_tv->uses();
+      if (uses_of_upcast_out_tv.size() < 2) {
+        continue;
+      }
+
+      // If this is the first use of the upcast output, keep it as is
+      if (expr == uses_of_upcast_out_tv.front()) {
+        continue;
+      }
+
+      auto upcast_out_tv_clone =
+          castOp(maybe_upcast_out_tv->dtype(), maybe_upcast_op->input(0));
+      std::cerr << "Recomputed cast: "
+                << upcast_out_tv_clone->definition()->toString();
+      expr = ir_utils::replaceValInExprInputs(
+          expr, maybe_upcast_out_tv, upcast_out_tv_clone);
+
+      privatized_upcast_ops_[maybe_upcast_op].insert(
+          upcast_out_tv_clone->definition()->as<UnaryOp>());
+    }
+  }
+}
+
+void SegmentCandidateFinder::revertPrivatizedUpcast(SegmentedGroup* group) {
+  for (const auto& [original_upcast, clones] : privatized_upcast_ops_) {
+    std::vector<UnaryOp*> upcast_in_group;
+    Val* upcast_val_to_keep = nullptr;
+    for (auto uop : ir_utils::filterByType<UnaryOp>(group->exprs())) {
+      if (uop != original_upcast && !clones.count(uop)) {
+        continue;
+      }
+
+      auto upcast_tv = uop->out();
+
+      // Prefer the original upcast if found
+      if (upcast_val_to_keep == nullptr ||
+          upcast_tv == original_upcast->out()) {
+        upcast_val_to_keep = upcast_tv;
+      }
+    }
+
+    if (upcast_in_group.size() < 2) {
+      continue;
+    }
+
+    for (const auto i : c10::irange(upcast_in_group.size())) {
+      UnaryOp* uop = upcast_in_group.at(i);
+      Val* upcast_val_to_replace = uop->out();
+      if (upcast_val_to_replace == upcast_val_to_keep) {
+        // Keep this uop as is since its output replaces the other
+        // upcast outputs
+        continue;
+      }
+
+      // Replace upcast_val_to_replace with upcast_val_to_keep
+      for (auto& expr : group->exprs_) {
+        auto input_it = std::find(
+            expr->inputs().begin(),
+            expr->inputs().end(),
+            upcast_val_to_replace);
+        if (input_it == expr->inputs().end()) {
+          continue;
+        }
+
+        auto updated_expr = ir_utils::replaceValInExprInputs(
+            expr, *input_it, upcast_val_to_keep);
+        expr = updated_expr;
+
+        // Update consumer edge vals
+        for (auto consumer_edge : group->consumer_edges) {
+          if (consumer_edge->val == upcast_val_to_replace) {
+            consumer_edge->val = upcast_val_to_keep;
+          }
+        }
+
+        // Note that it should not be necessary to do anything with
+        // group->output_vals since the inserted upcast ops should never produce
+        // fusion outputs.
```

Comment:
are we also expecting the `upcast_val_to_replace` shouldn't be a segmenter output neither?

Reply:
IIRC, segmenter outputs are represented as the vals of consumer edges, so updating them should be sufficient.